process/track/blockNotarizer.go

```diff
@@ -0,0 +1,228 @@
+package track
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+type blockNotarizer struct {
+	hasher      hashing.Hasher
+	marshalizer marshal.Marshalizer
+
+	startHeaders map[uint32]data.HeaderHandler
+
+	mutNotarizedHeaders sync.RWMutex
+	notarizedHeaders    map[uint32][]*headerInfo
+}
+
+// NewBlockNotarizer creates a block notarizer object which implements blockNotarizerHandler interface
+func NewBlockNotarizer(hasher hashing.Hasher, marshalizer marshal.Marshalizer) (*blockNotarizer, error) {
+	if check.IfNil(hasher) {
+		return nil, process.ErrNilHasher
+	}
+	if check.IfNil(marshalizer) {
+		return nil, process.ErrNilMarshalizer
+	}
+
+	bn := blockNotarizer{
+		hasher:      hasher,
+		marshalizer: marshalizer,
+	}
+
+	bn.notarizedHeaders = make(map[uint32][]*headerInfo)
+
+	return &bn, nil
+}
+
+func (bn *blockNotarizer) addNotarizedHeader(
+	shardID uint32,
+	notarizedHeader data.HeaderHandler,
+	notarizedHeaderHash []byte,
+) {
+	if check.IfNil(notarizedHeader) {
+		return
+	}
+
+	bn.mutNotarizedHeaders.Lock()
+	bn.notarizedHeaders[shardID] = append(bn.notarizedHeaders[shardID], &headerInfo{header: notarizedHeader, hash: notarizedHeaderHash})
+	if len(bn.notarizedHeaders[shardID]) > 1 {
+		sort.Slice(bn.notarizedHeaders[shardID], func(i, j int) bool {
+			return bn.notarizedHeaders[shardID][i].header.GetNonce() < bn.notarizedHeaders[shardID][j].header.GetNonce()
+		})
+	}
+	bn.mutNotarizedHeaders.Unlock()
+}
+
+func (bn *blockNotarizer) cleanupNotarizedHeadersBehindNonce(shardID uint32, nonce uint64) {
+	if nonce == 0 {
+		return
+	}
+
+	bn.mutNotarizedHeaders.Lock()
+	defer bn.mutNotarizedHeaders.Unlock()
+
+	notarizedHeaders, ok := bn.notarizedHeaders[shardID]
+	if !ok {
+		return
+	}
+
+	headersInfo := make([]*headerInfo, 0)
+	for _, headerInfo := range notarizedHeaders {
+		if headerInfo.header.GetNonce() < nonce {
+			continue
+		}
+
+		headersInfo = append(headersInfo, headerInfo)
+	}
+
+	bn.notarizedHeaders[shardID] = headersInfo
+}
+
+func (bn *blockNotarizer) displayNotarizedHeaders(shardID uint32, message string) {
```

Comment:
this should be class independet. you could have a display.go file and have all the functions there. Here the displayNotarizedHeaders should get the list / sorted notarized headers.

Reply:
Disagree, it depends on this structure as it operates on the same mutex with the rest of the functions. It is better to keep all functionality regarding the same data structures close together (in the same struct) [encapsulation] rather than have separate "structs" or files that groups functions based on behavior. Polymorphism also relies on the fact that multiple structs (classes) have the same function signatures.

Reply:
Actually display.go has been deleted and its functionality has been moved here and with this approach the duplicate code has been also removed.